### PR TITLE
fix for custom pipeline configuration 

### LINKED ISSUE #28907: wrongly annotated configuration when saving a model that has a custom pipeline
### System Info

- `transformers` version: 4.35.2
- Platform: Linux-6.1.58+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.1.0+cu121 (False)
- Tensorflow version (GPU?): 2.15.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.8.0 (cpu)
- Jax version: 0.4.23
- JaxLib version: 0.4.23
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@Narsil

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
# repo with custom pipeline
from transformers import AutoModelForImageClassification
model = AutoModelForImageClassification.from_pretrained("not-lain/MyRepo", trust_remote_code=True) 
```

```python
model.push_to_hub("testpushfrommodel")
```

```
.
└── testpushfrommodel
    ├── config.json
    └── model.safetensors
```

```json
{
  "_name_or_path": "not-lain/MyRepo",
  "architectures": [
    "MnistModel"
  ],
  "auto_map": {
    "AutoConfig": "not-lain/MyRepo--MyConfig.MnistConfig",
    "AutoModelForImageClassification": "not-lain/MyRepo--MyModel.MnistModel"
  },
  "conv1": 10,
  "conv2": 20,
  "custom_pipelines": {
    "image-classification": {
      "impl": "MyPipe.MnistPipe",
      "pt": [
        "AutoModelForImageClassification"
      ],
      "tf": [],
      "type": "image"
    }
  },
  "model_type": "MobileNetV1",
  "torch_dtype": "float32",
  "transformers_version": "4.35.2"
}
```

`impl` is wrongly annotated

### Expected behavior

output configuration is 
```json
{
  "_name_or_path": "not-lain/MyRepo",
  "architectures": [
    "MnistModel"
  ],
  "auto_map": {
    "AutoConfig": "not-lain/MyRepo--MyConfig.MnistConfig",
    "AutoModelForImageClassification": "not-lain/MyRepo--MyModel.MnistModel"
  },
  "conv1": 10,
  "conv2": 20,
  "custom_pipelines": {
    "image-classification": {
      "impl": "not-lain/MyRepo--MyPipe.MnistPipe",
      "pt": [
        "AutoModelForImageClassification"
      ],
      "tf": [],
      "type": "image"
    }
  },
  "model_type": "MobileNetV1",
  "torch_dtype": "float32",
  "transformers_version": "4.35.2"
}

### PR DESCRIPTION
**Pull request by not-lain**
# What does this PR do?
 
fixes configuration file not pointing at a remote repo with custom pipeline architecture

Fixes #28907 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@Rocketknight1

### CODE REVIEW

**Comment by not-lain** [comment]
@Rocketknight1 could you review this one too ?
fixed the tests (i forgor to update my branch :D )

**Comment by Rocketknight1** [comment]
Sure, I'll try to take a look at this one and the pipeline upload one!

**Comment by not-lain** [comment]
cc @Rocketknight1 any review on this one

**Review by Rocketknight1 — APPROVED** [review]
I reviewed this and I'm going to approve it on the basis that I think it fixes some specific issues and doesn't break anything that currently works, but I think we need much more clarity in the docs and our conceptual model of custom pipelines in future - this was quite a hard PR to review because I wasn't really sure what behaviour our library was "supposed" to have in these cases.

See my comment [here](https://github.com/huggingface/transformers/issues/28907#issuecomment-1959657870) for more.

**Comment by Rocketknight1** [comment]
cc @ArthurZucker for core maintainer review

**Comment by not-lain** [comment]
cc @Rocketknight1 @ArthurZucker 
I made this video to explain why this is important https://youtu.be/2LPJ_1QuK90

**Comment by not-lain** [comment]
it also explains the fundamentals of the transformers library on how to load the custom architecture from a repo that has the architecture defined on another one, which is a workaround i used before on [this](https://huggingface.co/briaai/RMBG-1.4/discussions/9) .

it's pretty much always depends on the config.json which is why it's essential to fix it

File: src/transformers/feature_extraction_utils.py
Comment (ArthurZucker) [inline src/transformers/feature_extraction_utils.py:541]: just curious as to why we only do it if not local whereas before we did it for local and not local? 🤗

**Review by ArthurZucker — COMMENTED** [review]
My main question is why don't we update the [`save_pretrained`](https://github.com/huggingface/transformers/blob/d03de1c1070cee0fb7b9358669abc3f68ef56b75/src/transformers/pipelines/base.py#L922) part as well to make sure what we already have the correct module? 
We can keep your changes for the previous checkpoints, and make sure new ones have the full path (like it is the case for tokenizer and model code path)
Also let's maybe add a test close to `https://github.com/huggingface/transformers/blob/d03de1c1070cee0fb7b9358669abc3f68ef56b75/tests/pipelines/test_pipelines_common.py#L709` to make sure we can load a custom pipeline from `not-laion` when it was pushed from `laion`!

File: src/transformers/feature_extraction_utils.py
Comment (not-lain) [reply src/transformers/feature_extraction_utils.py:541]: there is a nested if statements in there 
-if
--- if
--- if

Comment (not-lain) [reply src/transformers/feature_extraction_utils.py:541]: > what does `is_local` mean ? 

if i call a model using this 
```python
path = "./folder1/folder2/folder3"
from transformers import AutoModelForImageClassification
model = AutoModelForImageClassification.from_pretrained(path, trust_remote_code=True)
```
if the path exists in my pc then we load from there 
> what if the `is_local` is false ? 

then in this case we call the model from the hub and add the tag
> is the tag added when we are only calling a custom model from huggingface ?

yes, there is no need to add the tag if we are calling the model from the local pc

hope this answers your questions 🤗

**Comment by not-lain** [comment]
@ArthurZucker 
your confusion is on me, i have wrongly mentioned the issue, now after checking the code and experimenting more i now know it's when we are calling the model and not when we are saving/pushing.

again apologies for the confusion

**Review by ArthurZucker — COMMENTED** [review]
Let's wait for the test to do another round of review!

**Comment by not-lain** [comment]
cc @ArthurZucker
this should be ready for review

**Comment by not-lain** [comment]
@ArthurZucker @Rocketknight1 
could you review https://github.com/huggingface/transformers/pull/29172 first, since i need the push_to_hub feature from there added to the main branch to add more coverage to the tests just to be safe.

i'll update the tests on this pull request right after, sending lots of hugs 🤗🤗

**Comment by ArthurZucker** [comment]
Sorry on our late answers here! I was off for a while !

**Comment by not-lain** [comment]
@ArthurZucker hope you had fun ✨✨, anyways I have updated the tests to be more straight forward and to the point, do let me know if you have any reviews on this

**Comment by not-lain** [comment]
@ArthurZucker there's no need to wait for `29172` anymore since in 323b50d666a1fcb5ca09db7c6c4152b9bd5834a2 I switched to using the model to push to another repo (case finetuned model with custom pipeline), also friendly pinging for a review

**Comment by not-lain** [comment]
Hi @ArthurZucker 
Any reviews on this PR?

**Comment by not-lain** [comment]
hey @ArthurZucker , hope you're doing well.
can I get a review on this PR ?

**Comment by ArthurZucker** [comment]
Wow I am sorry @not-lain really my bad here

**Review by ArthurZucker — APPROVED** [review]
This seems to work well, I think overall I just don't like that we have to duplicate the logic everywhere but that's not on you! 
Thanks for your continuous contribution to the library and the ecosytem! And sorry for being so late here

**Comment by not-lain** [comment]
@ArthurZucker it's ok, you have been busy with lots of stuff including llama-3 and I know things are hectic on your side, just remember to take a break once in a while.
Also thanks a lot for the review, and I wish you a wonderful day ✨

**Pull request merged**
